### PR TITLE
Fabio

### PR DESCRIPTION
**Pull request by faguccio**
Exercise 3 solution

### CODE REVIEW

File: homework-3/student_submission.cpp [hunk 1]
```diff
@@ -12,17 +12,23 @@
 
 #define MEASURE_TIME true
 #define NUM_WORKERS 32
+#define NUM_GENERATORS 1
```
Comment (Davaff) [inline homework-3/student_submission.cpp:15]: Increase number of generators

File: homework-3/student_submission.cpp [hunk 2]
```diff
@@ -65,20 +71,30 @@ ProblemQueue problemQueue;
 // generate numProblems sha1 hashes with leadingZerosProblem leading zero bits
 // This method is intentionally compute intense so you can already start working on solving
 // problems while more problems are generated
-void generateProblem(int seed, int numProblems, int leadingZerosProblem){
-    generating = true;
-    srand(seed+1);
+void generateProblem(std::string problemRand[], int numProblems, int leadingZerosProblem, int idx){
+    // std::cout << numProblems << "\n";
```
Comment (Davaff) [inline homework-3/student_submission.cpp:75]: Remove commented in code

File: homework-3/student_submission.cpp [hunk 3]
```diff
@@ -65,20 +71,30 @@ ProblemQueue problemQueue;
...
+    Sha1Hash mock;
 
-    for(int i = 0; i < numProblems; i++){
-        std::string base = std::to_string(rand()) + std::to_string(rand());
-        Sha1Hash hash = Utility::sha1(base);
+    int step = numProblems / NUM_GENERATORS;
+    if (! (numProblems % NUM_GENERATORS == 0)) {std::cout << "ass error";exit(1);}
+
+    for(int i = idx*step; i < (idx+1)*step; i++){
```
Comment (Davaff) [inline homework-3/student_submission.cpp:81]: Create a thread pool instead of splitting the work here

**Review by Davaff — CHANGES_REQUESTED** [review]

File: homework-3/student_submission.cpp (hunk 3 above)
Comment (faguccio) [reply homework-3/student_submission.cpp:81]: That's not needed, the worker threads never extinguish the thread pool while the generator are working as we have way more workers than generators.

File: homework-3/student_submission.cpp [hunk 4]
```diff
@@ -65,20 +71,30 @@ ProblemQueue problemQueue;
...
         do{
             // we keep hashing ourself until we find the desired amount of leading zeros
             hash = Utility::sha1(hash);
         }while(Utility::count_leading_zero_bits(hash) < leadingZerosProblem);
         problemQueue.push(Problem{hash, i});
     }
-    generating = false;
+
+    finished_generator++;
+    // std::cout << "GENERATOR FINISHED\n";
```
Comment (PietroCaforio) [inline homework-3/student_submission.cpp:92]: Remove commented code

**Review by PietroCaforio — APPROVED** [review]

File: homework-3/student_submission.cpp (hunk 3 above)
Comment (Davaff) [reply homework-3/student_submission.cpp:81]: >  the worker threads never extinguish the thread pool 

Please clarify this sentence


> we have way more workers than generators

What if I want to change the number of generators and workers to try how it affects performance

Comment (faguccio) [reply homework-3/student_submission.cpp:81]: > What if I want to change the number of generators and workers to try how it affects performance

That's out of the scope of the project.

**Review by Davaff — APPROVED** [review]

**Pull request merged**